csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+private import semmle.code.csharp.dataflow.internal.DataFlowPublic
+private import semmle.code.csharp.frameworks.system.Collections
+private import semmle.code.csharp.frameworks.system.collections.Generic
+private import semmle.code.csharp.frameworks.system.linq.Expressions
+
+/** An unbound callable. */
+class SourceDeclarationCallable extends Callable {
+  SourceDeclarationCallable() { this = this.getSourceDeclaration() }
+}
+
+/** An unbound method. */
+class SourceDeclarationMethod extends SourceDeclarationCallable, Method { }
+
+private predicate hasDelegateArgumentPosition(SourceDeclarationCallable c, int i) {
+  exists(DelegateType dt |
+    dt = c.getParameter(i).getType().(SystemLinqExpressions::DelegateExtType).getDelegateType()
+  |
+    not dt.getReturnType() instanceof VoidType
+  )
+}
+
+private predicate hasDelegateArgumentPosition2(SourceDeclarationCallable c, int i, int j) {
+  exists(DelegateType dt |
+    dt = c.getParameter(i).getType().(SystemLinqExpressions::DelegateExtType).getDelegateType()
+  |
+    exists(dt.getParameter(j))
+  )
+}
+
+/** INTERNAL: Do not use. */
+module Internal {
+  newtype TContentStack =
+    TNilContentStack() or
+    TPushContentStack(Content head, ContentStack tail) {
+      tail = TNilContentStack()
+      or
+      exists(FrameworkDataFlow fdf |
+        fdf.requiresContentStack(head, tail) and
+        tail.length() < accessPathLimit()
+      )
+      or
+      tail = ContentStack::singleton(_) and
+      head instanceof ElementContent
+      or
+      tail = ContentStack::element()
+    }
+
+  newtype TSummaryInput =
+    TQualifierSummaryInput() or
+    TArgumentSummaryInput(int i) { i = any(Parameter p).getPosition() } or
+    TDelegateSummaryInput(int i) { hasDelegateArgumentPosition(_, i) }
+
+  newtype TSummaryOutput =
+    TQualifierSummaryOutput() or
+    TReturnSummaryOutput() or
+    TArgumentSummaryOutput(int i) {
+      exists(SourceDeclarationCallable c | exists(c.getParameter(i)))
+    } or
+    TDelegateSummaryOutput(int i, int j) { hasDelegateArgumentPosition2(_, i, j) }
+
+  newtype TFlowSource =
+    TRemoteFlowSource() or
+    TLocalFlowSource() or
+    TStoredFlowSource()
+
+  newtype TFlowSink =
+    TRemoteFlowSink() or
+    THtmlFlowSink() or
+    TEmailFlowSink() or
+    TSqlFlowSink() or
+    TCommandFlowSink()
+
+  /** Holds if a flow-source of kind `a` is also a flow-source of kind `b`. */
+  predicate flowSourceSubset(FlowSource a, FlowSource b) { none() }
+
+  /** Holds if a flow-sink of kind `a` is also a flow-sink of kind `b`. */
+  predicate flowSinkSubset(FlowSink a, FlowSink b) {
+    a in [THtmlFlowSink().(FlowSink), TEmailFlowSink().(FlowSink)] and
+    b = TRemoteFlowSink()
+  }
+}
+
+private import Internal
+
+/** A content stack. */
+class ContentStack extends TContentStack {
+  /** Gets the content stack obtained by popping `c` from this stack, if any. */
+  ContentStack pop(Content c) { this = TPushContentStack(c, result) }
```

Comment:
Since these are functional-style lists, perhaps `push` -> `cons` and `pop` -> `tail` throughout?

Reply:
Yeah, they were originally calls `AccessPath`s, but I wanted to call them something else to avoid confusion with the access paths from the shared library. I think I am just gonna go for `ContentList` instead.